Marlin/src/HAL/STM32/tft/tft_spi.cpp

```diff
@@ -34,43 +34,177 @@
+{
+  if ( speed >= (spibasefreq / 2) )
+  {
+    return SPI_BAUDRATEPRESCALER_2;
+  }
+  else if ( speed >= (spibasefreq / 4) )
+  {
+    return SPI_BAUDRATEPRESCALER_4;
+  }
+  else if ( speed >= (spibasefreq / 8) )
+  {
+    return SPI_BAUDRATEPRESCALER_8;
+  }
+  else if ( speed >= (spibasefreq / 16) )
+  {
+    return SPI_BAUDRATEPRESCALER_16;
+  }
+  else if ( speed >= (spibasefreq / 32) )
+  {
+    return SPI_BAUDRATEPRESCALER_32;
+  }
+  else if ( speed >= (spibasefreq / 64) )
+  {
+    return SPI_BAUDRATEPRESCALER_64;
+  }
+  else if ( speed >= (spibasefreq / 128) )
+  {
+    return SPI_BAUDRATEPRESCALER_128;
+  }
+  else
+  {
+    return SPI_BAUDRATEPRESCALER_256;
+  }
+}
+
+extern "C" {
+  #include <utility/spi_com.h>
+}
+
 void TFT_SPI::Init() {
+
   SPI_TypeDef *spiInstance;
 
   OUT_WRITE(TFT_A0_PIN, HIGH);
   OUT_WRITE(TFT_CS_PIN, HIGH);
 
   if ((spiInstance = (SPI_TypeDef *)pinmap_peripheral(digitalPinToPinName(TFT_SCK_PIN),  PinMap_SPI_SCLK)) == NP) return;
+
   if (spiInstance != (SPI_TypeDef *)pinmap_peripheral(digitalPinToPinName(TFT_MOSI_PIN), PinMap_SPI_MOSI)) return;
 
   #if PIN_EXISTS(TFT_MISO) && TFT_MISO_PIN != TFT_MOSI_PIN
     if (spiInstance != (SPI_TypeDef *)pinmap_peripheral(digitalPinToPinName(TFT_MISO_PIN), PinMap_SPI_MISO)) return;
   #endif
 
+  // SPI can have different baudrates depending on read or write functionality.
+
+#if PIN_EXISTS(TFT_MISO)
+  // static clkdiv_read variable.
+  bool has_clkdiv_read = false;
+#endif
+  // static clkdiv_write variable.
+  bool has_clkdiv_write = false;
+
+#if (PIN_EXISTS(TFT_MISO) && defined(TFT_BAUDRATE_READ)) || defined(TFT_BAURDATE_WRITE)
+    spi_t tmp_spi;
+    tmp_spi.pin_sclk = digitalPinToPinName(TFT_SCK_PIN);
+    uint32_t spibasefreq = spi_getClkFreq(&tmp_spi);
+
+#if PIN_EXISTS(TFT_MISO) && TFT_BAUDRATE_READ
+  clkdiv_read = _GetClockDivider(spibasefreq, TFT_BAUDRATE_READ);
+  has_clkdiv_read = true;
+#endif
+
+#ifdef TFT_BAUDRATE_WRITE
+  clkdiv_write = _GetClockDivider(spibasefreq, TFT_BAUDRATE_WRITE);
+  has_clkdiv_write = true;
+#endif
+
+#endif
+
+  if ( !has_clkdiv_write )
+  {
+#ifdef SPI1_BASE
+    if (spiInstance == SPI1)
+    {
+      clkdiv_write = SPI_BAUDRATEPRESCALER_4;
+      has_clkdiv_write = true;
+    }
+#endif
+
+    if ( !has_clkdiv_write )
+    {
+      clkdiv_write = SPI_BAUDRATEPRESCALER_2;
+    }
+  }
+
+#if PIN_EXISTS(TFT_MISO)
+  if ( !has_clkdiv_read )
+  {
+    clkdiv_read = clkdiv_write;
+  }
+#endif
+
   SPIx.Instance                = spiInstance;
   SPIx.State                   = HAL_SPI_STATE_RESET;
   SPIx.Init.NSS                = SPI_NSS_SOFT;
   SPIx.Init.Mode               = SPI_MODE_MASTER;
-  SPIx.Init.Direction          = (TFT_MISO_PIN == TFT_MOSI_PIN) ? SPI_DIRECTION_1LINE : SPI_DIRECTION_2LINES;
-  SPIx.Init.BaudRatePrescaler  = SPI_BAUDRATEPRESCALER_2;
+  SPIx.Init.Direction          = (TFT_MISO_PIN != NC && TFT_MISO_PIN == TFT_MOSI_PIN) ? SPI_DIRECTION_1LINE : SPI_DIRECTION_2LINES;
```

Comment:
This code is a logical error and a result is misunderstanding what SPIx.Init.Direction is.
SPI_DIRECTION_1LINE means that the same pin is used for sending and receiving data. Pin's role (MOSI/MISO) is set by MCU on HW level depending current SPI operation.
SPI_DIRECTION_2LINES means that pins roles are fixed. It does not mean that you need both MOSI and MISO pins.
When you use write-only or read only SPI devices from MCU's perspective this is SPI_DIRECTION_2LINES communication.
SPI_DIRECTION_1LINE should only be used when same pin is used for both sending and receiving data.
Original code had correct implementation.